src/fsharp/Optimizer.fs

```diff
@@ -2968,6 +2970,48 @@ and OptimizeLambdas (vspec: Val option) cenv env topValInfo e ety =
                  HasEffect=false
                  MightMakeCriticalTailcall = false
                  Info= valu }
+
+    // For partial applications, try to wrap it in a lambda and re-optimize as lambdas have better optimization strategies.
+    | Expr.App (Expr.Val (vref, _, _) as vexpr, fty, tinst, args, mApp) when isFunTy g ety ->
+        match vref.ValReprInfo, vspec with
+        | Some valInfo, Some v ->
+        //    let argTys, retTy = stripFunTy g ety
+
+            // We check the number of curried arguments with the applied arguments to ensure that it's safe to turn
+            //     this app expr into a lambda one.
+            // Example:
+            (*
+                let test a b c d e f g : int -> unit =
+                    printfn "hello"
+                    System.Console.WriteLine
+
+                let test2 a b c d e f g xs =
+                    let action = test a b c d e f g
+                    printfn "world"
+                    xs |> Array.iter action
+            *)
+            // In the example, we should not make a lambda around "test" in "test2" because "test" is fully applied but returns
+            //     a partial application.
+            if valInfo.NumCurriedArgs <> 0 && not v.IsMutable && not vref.IsMutable &&
```

Comment:
valInfo.NumCurriedArgs is used twice and can be put into a variable

Reply:
Most likely CSE will take care of that at runtime, forced locals aren't free either.